pkg/iac/terraform/state/backend/tfcloud_reader.go

```diff
@@ -28,72 +31,93 @@ type TFCloudBody struct {
 }
 
 type TFCloudBackend struct {
-	request *http.Request
-	client  pkghttp.HTTPClient
-	reader  io.ReadCloser
-	opts    *Options
+	client        *tfe.Client
+	reader        io.ReadCloser
+	opts          *Options
+	workspacePath string
 }
 
-func NewTFCloudReader(client pkghttp.HTTPClient, workspaceId string, opts *Options) (*TFCloudBackend, error) {
-	req, err := http.NewRequest(http.MethodGet, fmt.Sprintf("%s/workspaces/%s/current-state-version", opts.TFCloudEndpoint, workspaceId), nil)
+func NewTFCloudReader(client pkghttp.HTTPClient, workspacePath string, opts *Options) (*TFCloudBackend, error) {
+	token, err := GetToken(opts)
 	if err != nil {
 		return nil, err
 	}
-	req.Header.Add("Content-Type", "application/vnd.api+json")
-	return &TFCloudBackend{req, client, nil, opts}, nil
+	config := &tfe.Config{
+		Token:      token,
+		Address:    opts.TFCloudEndpoint,
+		HTTPClient: client.(*http.Client),
+	}
+	tfcClient, err := tfe.NewClient(config)
+	if err != nil {
+		return nil, err
+	}
+	return &TFCloudBackend{tfcClient, nil, opts, workspacePath}, nil
 }
 
-func (t *TFCloudBackend) authorize() error {
-	token := t.opts.TFCloudToken
+func GetToken(opts *Options) (string, error) {
+	token := opts.TFCloudToken
 	if token == "" {
 		tfConfigFile, err := getTerraformConfigFile()
 		if err != nil {
-			return err
+			return "", err
 		}
+
 		file, err := os.Open(tfConfigFile)
 		if err != nil {
-			return err
+			return "", err
 		}
 		defer file.Close()
 		reader := NewTFCloudConfigReader(file)
-		token, err = reader.GetToken(t.request.URL.Host)
+
+		u, err := url.Parse(opts.TFCloudEndpoint)
 		if err != nil {
-			return err
+			return "", err
 		}
+		return reader.GetToken(u.Host)
+	}
+	return token, nil
+}
+
+// A regular expression used to validate string workspace ID patterns.
+var reStringID = regexp.MustCompile(`^ws-[a-zA-Z0-9\-\._]+$`)
```

Comment:
Are we sure of this regex ? I can't find any place where this is documented, how do you managed to build this one ?

Reply:
Combination of [this](https://github.com/hashicorp/go-tfe/blob/main/workspace.go#L441) and following pattern that TFE has set in all workspaces. I don't think this is documented anywhere.

Reply:
I'm fine if this came from the tfe client 👍🏻 .